java-manta-client-unshaded/src/main/java/com/joyent/manta/client/MantaObjectOutputStream.java

```diff
@@ -170,54 +177,63 @@ public Thread newThread(final Runnable runnable) {
          * We have to wait here until the upload to Manta starts and a Writer
          * becomes available.
          */
-        while (httpContent.getWriter() == null) {
-            try {
-                Thread.sleep(CLOSED_CHECK_INTERVAL);
-            } catch (InterruptedException e) {
-                return;
+        synchronized (this.httpContent) {
+            while (!isClosed() && !httpContent.isClosed() && httpContent.getWriter() == null) {
+                try {
+                    /* We poll because the httpContent.notify() call within the
+                     * httpContent.writeTo() method may have been called before
+                     * the httpContent.wait() call below. */
+                    this.httpContent.wait(CLOSED_CHECK_INTERVAL);
+                } catch (InterruptedException e) {
+                    return;
+                }
```

Comment:
I admit it's been along time since I've used wait/notify and am having trouble seeing how this fits together.

Is this an optimization so we sleep less, or is the wait/notify dance also needed for correctness?  If it's for correctness I'm not sure I understand enough to make a safe change to this code in the future, but I'm also not sure what piece of understanding I'm missing.